DataDog/dd-trace-rb
# Avoid tracking logins from Devise::Strategies::Rememberable

**What does this PR do?**
In Devise integration [`Devise::Strategies::Authenticatable#validate`](https://github.com/heartcombo/devise/blob/main/lib/devise/strategies/authenticatable.rb#L37) is patched to track login successes and failures.

This method is not only used by [`Devise::Strategies::DatabaseAuthenticatable`](https://github.com/heartcombo/devise/blob/main/lib/devise/strategies/database_authenticatable.rb#L13), but also by [`Devise::Strategies::Rememberable`](https://github.com/heartcombo/devise/blob/main/lib/devise/strategies/rememberable.rb#L29) that is used for signing in the user using a remember token cookie.

**Motivation:**
This should reduce the amount of false-positive Login Success events for Devise AppSec integration.

**Additional Notes:**
I added an alias method for `validate` to avoid calling `super` from `Rememberable` strategy, since the super method is the `validate` from prepended `AuthenticatablePatch`.

**How to test the change?**
I tested this in a simple rails application with devise.
To test that regular login tracking works it's enough to log in via the sign-in form and observe a `users.login.success` tag on a trace.

To test that logins from `Devise::Strategies::Rememberable` strategy are not tracked:
- Log in via the sign-in form with checked "Remember Me" checkbox
- Go to any page that requires user authentication and delete the session cookie from Developer Tools
- Reload the page, the user will still be logged in, and the session cookie should reappear
- Observe that the trace was not tagged with `users.login.success`
